posydon/popsyn/rate_calculation.py

```diff
@@ -689,20 +703,230 @@ def compute_merger_rate_density(self, w_ijk, z_event, observable='DCOs', sensiti
+        if f'{sensitivity}_sensitivity' not in os.listdir(GRB_dir):
+            os.makedirs(sensitivity_dir)
+
+        # index of all DCOs
+        n = self.df.shape[0]
+        index = np.arange(0, n)
+
+        # redshif of each time bin
+        z_birth = self.get_centers_redshift_bins()
+
+        # define interpolator
+        get_redshift_from_time = self.get_redshift_from_cosmic_time_interpolator()
+
+        # hashmap to store everythig
+        data = {'index_1': {}, 'z_grb_1': {}, 'weights_1': {},
+                'index_2': {}, 'z_grb_2': {}, 'weights_2': {}}
+
+        # loop over all redshift bins
+        for i in tqdm(range(len(z_birth))):
+            
+            for s, event in enumerate(['CC1', 'CC2']):
+                # compute the CC time of each compact object
+                t_CC = self.get_time_GRB(z_birth[i], event)
+
+                # intrinsic population
+                if sensitivity == 'infinite':
+                    
+                    # sort out system not emitting GRBs
+                    bool_GRB = np.logical_and(t_CC < cosmology.age(1e-08).value*0.9999999 , self.df[f"GRB{s+1}"])
+                    
+                    # if there are no system emitting any GRB, continue
+                    if len(index[bool_GRB]) == 0:
+                        data[f'index_{s+1}'][str(z_birth[i])] = np.array([])
+                        data[f'z_grb_{s+1}'][str(z_birth[i])] = np.array([])
+                        data[f'weights_{s+1}'][str(z_birth[i])] = np.array([])
+                        continue
+                
+                    # get GRB redshifts
+                    z_GRB = get_redshift_from_time(t_CC[bool_GRB])
+
+                    # set detection probabilities to 1
+                    p_det = np.ones(len(index[bool_GRB]))
+
+                    # store index and redshift of merger
+                    data[f'index_{s+1}'][str(z_birth[i])] = index[bool_GRB]
+                    data[f'z_grb_{s+1}'][str(z_birth[i])] = z_GRB
+
+                    # compute and store cosmological rate weights as in eq. B.8 in Bavera et al. (2020)
+                    z_b = np.ones(len(index[bool_GRB]))*z_birth[i]
+                    w_ijk = self.merger_rate_weight(z_b, z_GRB, p_det, index[bool_GRB])
+                    data[f'weights_{s+1}'][str(z_birth[i])] = w_ijk
+
+                else:
+                    raise ValueError('Unknnown sensitivity!')
+
+        # TODO: implemet the saving to h5 files
+        if extention == 'npz':
+            if self.verbose:
+                print('Formatting the data ....')
+            for s in [1,2]:
+                data_to_save = [[],[],[],[]]
+                for i, dict in enumerate([data[key] for key in data.keys() if f'{s}' in key]):
+                    for key in dict.keys():
+                        data_to_save[i].extend(dict[key].tolist())
+                        if i == 0:
+                            data_to_save[3].extend(np.ones(len(dict[key]))*float(key))
```

Comment:
I'd suggest to use `[-1]` instead of `[3]`, this would allow future extensions in a more easy way. (the same applies then to line 894)